### PR TITLE
ref(metrics_extraction): total_count helper function

### PR DESCRIPTION
**Pull request by armenzg**
Some minor typing changes and a helper total count function.

### CODE REVIEW

File: src/sentry/snuba/metrics/extraction.py [hunk 1]
```diff
@@ -81,7 +83,7 @@
 }
 
 # Maps from Discover's syntax to Relay rule condition operators.
-_SEARCH_TO_RELAY_OPERATORS: Dict[str, "CompareOp"] = {
+_SEARCH_TO_RELAY_OPERATORS: Dict[str, CompareOp] = {
```
Comment (armenzg) [inline src/sentry/snuba/metrics/extraction.py:86]: `pyupgrade` fixed this.

This type is defined in here: https://github.com/getsentry/sentry/blob/24e79ddb8236012698979a9375cccac7b00d9422/src/sentry/snuba/metrics/extraction.py#L163

File: src/sentry/snuba/metrics/extraction.py [hunk 2]
```diff
@@ -620,13 +611,37 @@ def apdex_tag_spec(project: Project, argument: Optional[str]) -> List[TagSpec]:
     ]
 
 
-# This is used to map a metric to a function which generates a specification
-_DERIVED_METRICS: Dict[MetricOperationType, TagsSpecsGenerator] = {
-    "on_demand_failure_count": failure_tag_spec,
-    "on_demand_failure_rate": failure_tag_spec,
-    "on_demand_apdex": apdex_tag_spec,
+DERIVED_METRICS = {
```
Comment (armenzg) [inline src/sentry/snuba/metrics/extraction.py:614]: This is the centralized place in this file to defined derived metrics.

File: src/sentry/snuba/metrics/fields/base.py [hunk 3]
```diff
@@ -1785,21 +1785,21 @@ def generate_where_statements(
         ),
         # Custom operations used for on demand derived metrics.
         DerivedOp(
-            op="on_demand_failure_count",
+            op="on_demand_apdex",
```
Comment (armenzg) [inline src/sentry/snuba/metrics/fields/base.py:1788]: I'm just sorting these in alphabetical order.

Ideally, we would define all these `DerivedOp` by iterating over `DERIVED_METRICS`.

File: src/sentry/snuba/metrics/fields/snql.py [hunk 4]
```diff
@@ -802,22 +802,28 @@ def max_timestamp(aggregate_filter, org_id, use_case_id, alias=None):
     return timestamp_column_snql("maxIf", aggregate_filter, org_id, use_case_id, alias)
 
 
-def on_demand_failure_rate_snql_factory(aggregate_filter, org_id, use_case_id, alias=None):
+def total_count(aggregate_filter: Function) -> Function:
+    return Function("sumIf", [Column("value"), aggregate_filter])
```
Comment (armenzg) [inline src/sentry/snuba/metrics/fields/snql.py:806]: We use this function in a couple of places, thus, we can use this helper function.

File: src/sentry/snuba/metrics/fields/snql.py [hunk 5]
```diff
@@ -842,7 +848,9 @@ def on_demand_failure_count_snql_factory(
     )
 
 
-def on_demand_apdex_snql_factory(aggregate_filter, org_id, use_case_id, alias=None):
+def on_demand_apdex_snql_factory(
+    aggregate_filter: Function, org_id: int, use_case_id: UseCaseID, alias: Optional[str]
```
Comment (armenzg) [inline src/sentry/snuba/metrics/fields/snql.py:852]: Do we need to keep `None` as the default value for `alias`?

Comment (evanh) [reply src/sentry/snuba/metrics/fields/snql.py:852]: What other default would you use?

**Review by obostjancic — APPROVED** [review]

File: src/sentry/snuba/metrics/extraction.py [hunk 6]
```diff
@@ -620,13 +611,37 @@ def apdex_tag_spec(project: Project, argument: Optional[str]) -> List[TagSpec]:
...
+    "failure_rate": {
+        "tag_spec_func": failure_tag_spec,
+        "hash_method": "op",
+    },
 }
 
+# This is used to map a metric to a function which generates a specification
+_DERIVED_METRICS: Dict[MetricOperationType, TagsSpecsGenerator] = {}
+for metric in DERIVED_METRICS.keys():
+    _DERIVED_METRICS[cast(MetricOperationType, f"on_demand_{metric}")] = cast(
```
Comment (asottile-sentry) [inline src/sentry/snuba/metrics/extraction.py:632]: cast removes all type safety and should really only be used for bugs in mypy -- these should go back to being static dictionaries and a test should be added to check they are in sync

same for the other usages of cast

File: src/sentry/snuba/metrics/extraction.py [hunk 7]
```diff
@@ -620,13 +611,37 @@ def apdex_tag_spec(project: Project, argument: Optional[str]) -> List[TagSpec]:
...
+    },
+    "failure_rate": {
+        "tag_spec_func": failure_tag_spec,
+        "hash_method": "op",
+    },
 }
 
+# This is used to map a metric to a function which generates a specification
+_DERIVED_METRICS: Dict[MetricOperationType, TagsSpecsGenerator] = {}
+for metric in DERIVED_METRICS.keys():
```
Comment (asottile-sentry) [inline src/sentry/snuba/metrics/extraction.py:631]: `.keys()` is almost never needed in python (unless you are doing setops on a dictionary's parts) -- iterating a dictionary iterates its keys already

**Review by asottile-sentry — DISMISSED** [review]

**Comment by armenzg** [comment]
I've reduced the scope of this PR.

Trying not to use `cast` is a lot of work (which defeats the purpose of typing).

**Pull request merged**